File: sdk/ts/src/alias.ts
/**
 * alias.ts — YAML-file-based command alias expansion.
 *
 * Port of Go's cli/alias package. Aliases live in per-tool YAML
 * config files with three-tier priority: seeded < global < local.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import { Command } from 'commander';

export type AliasMap = Record<string, string>;

export interface ExpanderConfig {
  globalPath: string;
  localPath: string;
  seededAliases?: AliasMap;
  builtins?: Set<string>;
}

// ── Low-level helpers ───────────────────────────────────────

/** Read aliases from a single YAML config file. */
export function loadFrom(filePath: string): AliasMap {
  let data: string;
  try {
    data = fs.readFileSync(filePath, 'utf-8');
  } catch {
    return {};
  }

  const raw = yaml.load(data) as Record<string, unknown> | null;
  if (!raw || typeof raw !== 'object' || !raw['aliases']) return {};
  return raw['aliases'] as AliasMap;
}

/** Write aliases into a YAML config file, preserving other keys. */
export function saveTo(filePath: string, aliases: AliasMap): void {
  fs.mkdirSync(path.dirname(filePath), { recursive: true, mode: 0o750 });

  let existing: Record<string, unknown> = {};
  try {
    const data = fs.readFileSync(filePath, 'utf-8');
    existing = (yaml.load(data) as Record<string, unknown>) ?? {};
  } catch {
    // file doesn't exist yet
  }

  if (Object.keys(aliases).length === 0) {
    delete existing['aliases'];
  } else {
    existing['aliases'] = { ...aliases };
  }

  fs.writeFileSync(filePath, yaml.dump(existing), { mode: 0o600 });
}

/**
 * Find the first non-flag element in a slice.
 * Returns [index, value]. Flags with `=` are self-contained;
 * short flags without `=` consume the next element as their value.
 */
export function findFirstNonFlag(
  slice: string[],
): [number, string] {
  for (let i = 0; i < slice.length; i++) {
    const a = slice[i];
    if (!a.startsWith('-')) return [i, a];
    if (a.includes('=')) continue;
    // short/long flag consumes next arg as its value
    if (i + 1 < slice.length && !slice[i + 1].startsWith('-')) {
      i++;
    }
  }
  return [-1, ''];
}

// ── AliasStore ──────────────────────────────────────────────

/**
 * Simple single-file alias store with load/save, set/get/remove,
 * and first-arg expansion.
 */
export class AliasStore {
  private filePath: string;
  private aliases: AliasMap = {};

  constructor(filePath: string) {
    this.filePath = filePath;
  }

  load(): void {
    this.aliases = loadFrom(this.filePath);
  }

  save(): void {
    saveTo(this.filePath, this.aliases);
  }

  set(name: string, target: string): void {
    this.aliases[name] = target;
  }

  remove(name: string): void {
    delete this.aliases[name];
  }

  get(name: string): string | undefined {
    return this.aliases[name];
  }

  /** Returns a shallow copy. */
  all(): AliasMap {
    return { ...this.aliases };
  }

  /** Expand first arg if it matches an alias. */
  expand(args: string[]): string[] {
    if (args.length < 2) return args;

    const candidate = args[1];
    const expansion = this.aliases[candidate];
    if (!expansion) return args;

    const parts = expansion.split(/\s+/);
    return [args[0], ...parts, ...args.slice(2)];
  }
}

// ── Expander (multi-file, three-tier) ───────────────────────

/**
 * Three-tier alias expander matching Go's Expander.
 * Priority: seeded < global < local.
 */
export class Expander {
  private cfg: ExpanderConfig;

  constructor(cfg: ExpanderConfig) {
    this.cfg = cfg;
  }

  /** Merged alias map (seeded < global < local). */
  load(): AliasMap {
    const merged: AliasMap = {};

    // seeded (lowest priority)
    if (this.cfg.seededAliases) {
      Object.assign(merged, this.cfg.seededAliases);
    }

    // global overrides seeded
    const global = loadFrom(this.cfg.globalPath);
    Object.assign(merged, global);

    // local overrides global
    const local = loadFrom(this.cfg.localPath);
    Object.assign(merged, local);

    return merged;
  }

  /**
   * Rewrite args if args[1] (or first non-flag arg) matches
   * an alias. Returns [rewrittenArgs, wasExpanded].
   */
  expand(args: string[]): [string[], boolean] {
    if (args.length < 2) return [args, false];

    const aliases = this.load();
    if (Object.keys(aliases).length === 0) return [args, false];

    const [firstNonFlag, candidate] = findFirstNonFlag(args.slice(1));
    if (candidate === '') return [args, false];

    const expansion = aliases[candidate];
    if (!expansion) return [args, false];

    const parts = expansion.split(/\s+/);
    const prefix = args.slice(1, 1 + firstNonFlag);
    const suffix = args.slice(1 + firstNonFlag + 1);

    const result = [args[0], ...prefix, ...parts, ...suffix];
    return [result, true];
  }

  /** Reject names that shadow builtins or contain whitespace. */
  validateName(name: string): void {
    if (name === '') {
      throw new Error('alias name must not be empty');
    }
    if (/\s/.test(name)) {
      throw new Error('alias name must not contain whitespace');
    }
    if (this.cfg.builtins?.has(name)) {
      throw new Error(`alias "${name}" conflicts with a built-in command`);
    }
  }
}

// ── Commander bridge ────────────────────────────────────────

/** Walk the Commander program tree to find a command by space-delimited path. */
export function findCommand(
  program: Command,
  cmdPath: string,
): Command | undefined {
  const parts = cmdPath.split(' ');
  let cmd: Command | undefined = program;
  for (const part of parts) {
    cmd = cmd?.commands.find(c => c.name() === part);
    if (!cmd) return undefined;
  }
  return cmd;
}

/**
 * Register loaded aliases with Commander's native .alias() API.
 *
 * Simple single-command aliases use Commander's built-in .alias(),
 * giving free completion and help display. Multi-word or unresolved
 * targets get a hidden proxy command that re-parses with expanded args.
 */
export function bridgeAliases(
  program: Command,
  store: AliasStore,
): void {
  const all = store.all();
  for (const [name, target] of Object.entries(all)) {
    const parts = target.split(/\s+/);
    // simple alias: single word target that exists as a direct subcommand
    if (parts.length === 1) {
      const targetCmd = findCommand(program, target);
      if (targetCmd) {
        targetCmd.alias(name);
        continue;
      }
    }

    // multi-word target or unknown command: hidden proxy
    const proxy = new Command(name)
      .description(`Alias for ${target}`)
      .allowUnknownOption(true)
      .allowExcessArguments(true)
      .action(function (this: Command) {
        const expanded = target.split(/\s+/);
        program.parse([
          'node', program.name(), ...expanded, ...this.args,
        ]);
      });
    proxy.helpOption(false);
    (proxy as any)._hidden = true;
    program.addCommand(proxy);
  }
}

File: sdk/py/hop_top_kit/alias.py
"""
hop_top_kit.alias — YAML-file-based command alias expansion.

Three-tier priority: seeded (built-in) < global (user) < local (project).

Usage::

    expander = Expander(Config(
        global_path="~/.config/mytool/config.yaml",
        local_path=".mytool/config.yaml",
        seeded_aliases={"setup": "config interactive"},
    ))
    args, expanded = expander.expand(sys.argv)
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field

import click
import yaml


@dataclass
class Config:
    """Configures the Expander."""

    global_path: str = ""
    local_path: str = ""
    seeded_aliases: dict[str, str] = field(default_factory=dict)
    builtins: set[str] = field(default_factory=set)


class Expander:
    """Loads and expands YAML-file-based command aliases."""

    def __init__(self, cfg: Config) -> None:
        self._cfg = cfg

    def load(self) -> dict[str, str]:
        """Return merged alias map (seeded < global < local)."""
        try:
            gl = load_from(self._cfg.global_path)
        except (FileNotFoundError, OSError):
            gl = {}

        try:
            loc = load_from(self._cfg.local_path)
        except (FileNotFoundError, OSError):
            loc = {}

        merged: dict[str, str] = {}
        merged.update(self._cfg.seeded_aliases)
        merged.update(gl)
        merged.update(loc)
        return merged

    def expand(self, args: list[str]) -> tuple[list[str], bool]:
        """Rewrite args if first non-flag arg matches an alias.

        Returns (possibly rewritten args, whether expansion occurred).
        """
        if len(args) < 2:
            return args, False

        aliases = self.load()
        if not aliases:
            return args, False

        idx, candidate = find_first_non_flag(args[1:])
        if not candidate:
            return args, False

        expansion = aliases.get(candidate)
        if expansion is None:
            return args, False

        parts = expansion.split()
        prefix = args[1 : 1 + idx]
        suffix = args[1 + idx + 1 :]
        result = [args[0], *prefix, *parts, *suffix]
        return result, True

    def validate_name(self, name: str) -> None:
        """Raise ValueError if name is invalid for an alias."""
        if not name:
            raise ValueError("alias name must not be empty")
        if any(c in name for c in (" ", "\t", "\n")):
            raise ValueError("alias name must not contain whitespace")
        if name in self._cfg.builtins:
            raise ValueError(f"alias {name!r} conflicts with a built-in command")


def load_from(path: str) -> dict[str, str]:
    """Read aliases from a single YAML config file."""
    with open(path) as f:
        raw = yaml.safe_load(f)
    if not isinstance(raw, dict):
        return {}
    return dict(raw.get("aliases") or {})


def save_to(path: str, aliases: dict[str, str]) -> None:
    """Write aliases into a YAML config, preserving other keys."""
    os.makedirs(os.path.dirname(path), exist_ok=True)

    existing: dict = {}
    try:
        with open(path) as f:
            existing = yaml.safe_load(f) or {}
    except FileNotFoundError:
        pass

    if aliases:
        existing["aliases"] = dict(aliases)
    else:
        existing.pop("aliases", None)

    with open(path, "w") as f:
        yaml.dump(existing, f, default_flow_style=False)


def bridge_to_click(group: click.Group, store_path: str) -> None:
    """Patch a Click group's get_command to resolve aliases.

    Uses Click's AliasedGroup pattern: overrides get_command() so
    aliases resolve natively, and list_commands() so aliases appear
    in shell completions for free.
    """
    try:
        aliases = load_from(store_path)
    except (FileNotFoundError, OSError):
        return
    if not aliases:
        return

    original_get_command = group.get_command
    original_list_commands = group.list_commands

    def get_command_with_aliases(
        ctx: click.Context,
        cmd_name: str,
    ) -> click.Command | None:
        # real command takes precedence
        rv = original_get_command(ctx, cmd_name)
        if rv is not None:
            return rv
        target = aliases.get(cmd_name)
        if not target:
            return None
        parts = target.split()
        return original_get_command(ctx, parts[0])

    def list_commands_with_aliases(ctx: click.Context) -> list[str]:
        cmds = original_list_commands(ctx)
        return sorted(set(cmds) | set(aliases.keys()))

    group.get_command = get_command_with_aliases  # type: ignore[assignment]
    group.list_commands = list_commands_with_aliases  # type: ignore[assignment]


def find_first_non_flag(
    slc: list[str],
) -> tuple[int, str]:
    """Index+value of first non-flag element in slice."""
    i = 0
    while i < len(slc):
        a = slc[i]
        if not a.startswith("-"):
            return i, a
        if "=" in a:
            i += 1
            continue
        # Long flags (--foo) must use --flag=value for values;
        # bare --flag is treated as boolean. Only short flags
        # (-x val) consume the next arg.
        if a.startswith("--"):
            i += 1
            continue
        # short flag without '=': next arg is flag value
        if i + 1 < len(slc) and not slc[i + 1].startswith("-"):
            i += 2
        else:
            i += 1
    return -1, ""
